File: comparisons/MinsC2Rust/c_algorithm/Output/test_project/tests/test_binary_heap.rs
use test_project::binary_heap::{
    binary_heap_cmp, binary_heap_free, binary_heap_insert, binary_heap_new,
    binary_heap_num_entries, binary_heap_pop, BinaryHeap, BinaryHeapType, BinaryHeapValue,
};

pub const NUM_TEST_VALUES: usize = 10000;
#[test]
pub fn test_binary_heap_insert() {
    let mut heap = binary_heap_new(BinaryHeapType::Min, int_compare).unwrap();
    let mut test_array: Vec<i32> = (0..NUM_TEST_VALUES as i32).collect();

    for i in 0..NUM_TEST_VALUES {
        assert!(binary_heap_insert(&mut heap, test_array[i]));
    }

    assert_eq!(binary_heap_num_entries(&heap), NUM_TEST_VALUES);

    binary_heap_free(heap);
}

#[test]
pub fn test_binary_heap_new_free() {
    for _ in 0..NUM_TEST_VALUES {
        let heap = binary_heap_new::<i32>(BinaryHeapType::Min, int_compare);
        binary_heap_free(heap.unwrap());
    }
}

#[test]
pub fn test_out_of_memory() {
    let mut heap = binary_heap_new(BinaryHeapType::Min, int_compare).unwrap();
    let values = vec![15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0];

    for i in 0..16 {
        assert!(binary_heap_insert(&mut heap, values[i]));
    }

    assert_eq!(binary_heap_num_entries(&heap), 16);

    for i in 0..16 {
        assert!(!binary_heap_insert(&mut heap, values[i]));
        assert_eq!(binary_heap_num_entries(&heap), 16);
    }

    for i in 0..16 {
        let value = binary_heap_pop(&mut heap).unwrap();
        assert_eq!(value, i);
    }

    assert_eq!(binary_heap_num_entries(&heap), 0);

    binary_heap_free(heap);
}

#[test]
pub fn test_min_heap() {
    let mut heap = binary_heap_new(BinaryHeapType::Min, int_compare).unwrap();
    let mut test_array: Vec<i32> = (0..NUM_TEST_VALUES as i32).collect();

    for i in 0..NUM_TEST_VALUES {
        assert!(binary_heap_insert(&mut heap, test_array[i]));
    }

    let mut i = -1;
    while binary_heap_num_entries(&heap) > 0 {
        let val = binary_heap_pop(&mut heap).unwrap();
        assert!(val == i + 1);
        i = val;
    }

    assert_eq!(binary_heap_num_entries(&heap), 0);
    assert!(binary_heap_pop(&mut heap).is_none());

    binary_heap_free(heap);
}

#[test]
pub fn test_max_heap() {
    pub const NUM_TEST_VALUES: usize = 10000;
    let mut test_array: Vec<i32> = Vec::with_capacity(NUM_TEST_VALUES);

    let mut heap = binary_heap_new(BinaryHeapType::Max, |a: &i32, b: &i32| {
        if a < b {
            -1
        } else if a > b {
            1
        } else {
            0
        }
    })
    .unwrap();

    for i in 0..NUM_TEST_VALUES {
        test_array.push(i as i32);
        assert!(binary_heap_insert(&mut heap, test_array[i]));
    }

    let mut i = NUM_TEST_VALUES as i32;
    while binary_heap_num_entries(&heap) > 0 {
        let val = binary_heap_pop(&mut heap).unwrap();
        assert_eq!(val, i - 1);
        i = val;
    }

    binary_heap_free(heap);
}

pub fn int_compare(a: &i32, b: &i32) -> i32 {
    if a < b {
        -1
    } else if a > b {
        1
    } else {
        0
    }
}

File: c-algorithm/test/test-binary-heap.c
// 识别到的包含指令
#include <stdlib.h>
#include <assert.h>
#include "alloc-testing.h"
#include "framework.h"

// 识别到的全局变量
_Atomic size_t num_assert = 0;

// 识别到的宏定义
#define assert(expr)                                                           \
  num_assert += 1;                                                             \
  ((void)sizeof((expr) ? 1 : 0), __extension__({                               \
     if (expr)                                                                 \
       ; /* empty */

// 识别到的包含指令
#include "binary-heap.h"
#include "compare-int.h"

// 识别到的宏定义
#define NUM_TEST_VALUES 10000

// 识别到的全局变量
int test_array[NUM_TEST_VALUES];

// 识别到的函数定义
// 函数: test_binary_heap_new_free (line 45)
void test_binary_heap_new_free(void)
{
	BinaryHeap *heap;
	int i;

	for (i=0; i<NUM_TEST_VALUES; ++i) {
		heap = binary_heap_new(BINARY_HEAP_TYPE_MIN, int_compare);
		binary_heap_free(heap);
	}

	/* Test low memory scenario */

	alloc_test_set_limit(0);
	heap = binary_heap_new(BINARY_HEAP_TYPE_MIN, int_compare);
	assert(heap == NULL);

	alloc_test_set_limit(1);
	heap = binary_heap_new(BINARY_HEAP_TYPE_MIN, int_compare);
	assert(heap == NULL);
}

// 函数: test_binary_heap_insert (line 66)
void test_binary_heap_insert(void)
{
	BinaryHeap *heap;
	int i;

	heap = binary_heap_new(BINARY_HEAP_TYPE_MIN, int_compare);

	for (i=0; i<NUM_TEST_VALUES; ++i) {
		test_array[i] = i;
		assert(binary_heap_insert(heap, &test_array[i]) != 0);
	}

	assert(binary_heap_num_entries(heap) == NUM_TEST_VALUES);

	binary_heap_free(heap);
}

// 函数: test_min_heap (line 83)
void test_min_heap(void)
{
	BinaryHeap *heap;
	int *val;
	int i;

	heap = binary_heap_new(BINARY_HEAP_TYPE_MIN, int_compare);

	/* Push a load of values onto the heap */

	for (i=0; i<NUM_TEST_VALUES; ++i) {
		test_array[i] = i;
		assert(binary_heap_insert(heap, &test_array[i]) != 0);
	}

	/* Pop values off the heap and check they are in order */

	i = -1;
	while (binary_heap_num_entries(heap) > 0) {
		val = (int *) binary_heap_pop(heap);

		assert(*val == i + 1);
		i = *val;
	}

	/* Test popping from an empty heap */

	assert(binary_heap_num_entries(heap) == 0);
	assert(binary_heap_pop(heap) == BINARY_HEAP_NULL);

	binary_heap_free(heap);
}

// 函数: test_max_heap (line 116)
void test_max_heap(void)
{
	BinaryHeap *heap;
	int *val;
	int i;

	heap = binary_heap_new(BINARY_HEAP_TYPE_MAX, int_compare);

	/* Push a load of values onto the heap */

	for (i=0; i<NUM_TEST_VALUES; ++i) {
		test_array[i] = i;
		assert(binary_heap_insert(heap, &test_array[i]) != 0);
	}

	/* Pop values off the heap and check they are in order */

	i = NUM_TEST_VALUES;
	while (binary_heap_num_entries(heap) > 0) {
		val = (int *) binary_heap_pop(heap);

		assert(*val == i - 1);
		i = *val;
	}

	binary_heap_free(heap);
}

// 函数: test_out_of_memory (line 146)
void test_out_of_memory(void)
{
	BinaryHeap *heap;
	int *value;
	int values[] = {
		15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0,
	};
	int i;

	/* Allocate a heap and fill to the default limit */

	heap = binary_heap_new(BINARY_HEAP_TYPE_MIN, int_compare);

	alloc_test_set_limit(0);

	for (i=0; i<16; ++i) {
		assert(binary_heap_insert(heap, &values[i]) != 0);
	}

	assert(binary_heap_num_entries(heap) == 16);

	/* Check that we cannot add new values */

	for (i=0; i<16; ++i) {
		assert(binary_heap_insert(heap, &values[i]) == 0);
		assert(binary_heap_num_entries(heap) == 16);
	}

	/* Check that we can read the values back out again and they
	 * are in the right order. */

	for (i=0; i<16; ++i) {
		value = binary_heap_pop(heap);
		assert(*value == i);
	}

	assert(binary_heap_num_entries(heap) == 0);

	binary_heap_free(heap);
}

// 识别到的全局变量
static UnitTestFunction tests[] = {
	test_binary_heap_new_free,
	test_binary_heap_insert,
	test_min_heap,
	test_max_heap,
	test_out_of_memory,
	NULL
};

// 识别到的函数定义
// 函数: main (line 196)
int main(int argc, char *argv[])
{
	run_tests(tests);
  printf("num_assert: %lu\n", num_assert);
	return 0;
}
